File: 3419.minimize-the-maximum-edge-weight-of-graph.py
#
# @lc app=leetcode id=3419 lang=python3
#
# [3419] Minimize the Maximum Edge Weight of Graph
#

# @lc code=start
import collections

class Solution:
    def minMaxWeight(self, n: int, edges: List[List[int]], threshold: int) -> int:
        # Build the reversed graph
        # rev_adj[v] contains (u, w) indicating an original edge u -> v with weight w
        rev_adj = [[] for _ in range(n)]
        unique_weights = set()
        
        for u, v, w in edges:
            rev_adj[v].append((u, w))
            unique_weights.add(w)
            
        sorted_weights = sorted(list(unique_weights))
        
        if not sorted_weights:
            return -1
            
        # Function to check if all nodes can reach node 0 (in original graph)
        # using only edges with weight <= limit.
        # This is equivalent to checking if node 0 can reach all nodes 
        # in the reversed graph.
        def check(limit):
            count = 1
            visited = [False] * n
            visited[0] = True
            queue = collections.deque([0])
            
            while queue:
                u = queue.popleft()
                for v, w in rev_adj[u]:
                    if w <= limit and not visited[v]:
                        visited[v] = True
                        count += 1
                        queue.append(v)
            return count == n

        # If even with the maximum available weight we cannot connect the graph, return -1
        if not check(sorted_weights[-1]):
            return -1
            
        # Binary Search for the minimum weight
        left, right = 0, len(sorted_weights) - 1
        ans = sorted_weights[-1]
        
        while left <= right:
            mid = (left + right) // 2
            val = sorted_weights[mid]
            if check(val):
                ans = val
                right = mid - 1
            else:
                left = mid + 1
                
        return ans
# @lc code=end

File: 3419.minimize-the-maximum-edge-weight-of-graph.cpp
#
# @lc app=leetcode id=3419 lang=cpp
#
# [3419] Minimize the Maximum Edge Weight of Graph
#
# @lc code=start
class Solution {
public:
    bool canReach(int n, vector<vector<int>>& edges, int maxWeight, int threshold) {
        vector<vector<pair<int, int>>> adj(n);
        
        for (const auto& e : edges) {
            int u = e[0], v = e[1], w = e[2];
            if (w <= maxWeight) {
                adj[u].push_back({v, w});
            }
        }
        
        for (int i = 0; i < n; i++) {
            if ((int)adj[i].size() > threshold) {
                sort(adj[i].begin(), adj[i].end(), 
                     [](const auto& a, const auto& b) {
                         return a.second < b.second;
                     });
                adj[i].resize(threshold);
            }
        }
        
        for (int start = 1; start < n; start++) {
            queue<int> q;
            vector<bool> visited(n, false);
            q.push(start);
            visited[start] = true;
            bool found = false;
            
            while (!q.empty()) {
                int u = q.front();
                q.pop();
                
                if (u == 0) {
                    found = true;
                    break;
                }
                
                for (const auto& [v, w] : adj[u]) {
                    if (!visited[v]) {
                        visited[v] = true;
                        q.push(v);
                    }
                }
            }
            
            if (!found) return false;
        }
        
        return true;
    }
    
    int minMaxWeight(int n, vector<vector<int>>& edges, int threshold) {
        if (edges.empty()) return -1;
        
        set<int> weights;
        for (const auto& e : edges) {
            weights.insert(e[2]);
        }
        
        vector<int> sortedWeights(weights.begin(), weights.end());
        
        int left = 0, right = sortedWeights.size() - 1;
        int result = -1;
        
        while (left <= right) {
            int mid = left + (right - left) / 2;
            int maxWeight = sortedWeights[mid];
            
            if (canReach(n, edges, maxWeight, threshold)) {
                result = maxWeight;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        
        return result;
    }
};
# @lc code=end